pkg/apis/operator/v1alpha1/type.go

```diff
@@ -0,0 +1,777 @@
+/*
+Copyright 2022 The Karmada Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:path=karmadadeployments,scope=Namespaced,categories={karmada-io}
+// +kubebuilder:printcolumn:JSONPath=`.status.controlPlaneReady`,name="Status",type=string
+// +kubebuilder:printcolumn:JSONPath=`.metadata.creationTimestamp`,name="Age",type=date
+
+// KarmadaDeployment enables declarative installation of karmada.
+type KarmadaDeployment struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired behavior of the KarmadaDeployment.
+	// +optional
+	Spec KarmadaDeploymentSpec `json:"spec,omitempty"`
+
+	// Most recently observed status of the KarmadaDeployment.
+	// +optional
+	Status KarmadaDeploymentStatus `json:"status,omitempty"`
+}
+
+// KarmadaDeploymentSpec is the specification of the desired behavior of the KarmadaDeployment.
+type KarmadaDeploymentSpec struct {
+	// HostCluster represents the cluster where to install the Karmada control plane.
+	// If not set, the control plane will be installed on the cluster where
+	// running the operator.
+	// +optional
+	HostCluster *HostCluster `json:"hostCluster,omitempty"`
+
+	// set global images repository and tag, kubenetes original images and
+	// karmada images can be set up separately. If there is nil, will use the
+	// default images to install.
+	// +optional
+	GlobalImage *GlobalImage `json:"globalImage,omitempty"`
+
+	// Networking holds configuration for the networking topology of the cluster.
+	// +optional
+	Networking *Networking `json:"networking,omitempty"`
+
+	// Etcd holds configuration for etcd.
+	// +required
+	Etcd *Etcd `json:"etcd,omitempty"`
+
+	// BaseComponents holds configuration for all karmada base components.
+	// +optional
+	BaseComponents *BaseComponents `json:"baseComponents,omitempty"`
+
+	// AdvanceComponents holds configuration for all karmada optional components.
+	// +optional
+	AdvanceComponents *AdvanceComponents `json:"advanceComponents,omitempty"`
```

Comment:
I think we can put all of them into a struct like:
```golang
Components KarmadaComponents `json:"components,omitempty"`
```
The reasons are:
1. ETCD is also a component 
2. we can't say exactly which component is belongs to basic(required) or advanced(optional), such as for some case, they don't want to deploy the karmada-apiserver and etcd.

Reply:
> 1. ETCD is also a component
> 2. we can't say exactly which component is belongs to basic(required) or advanced(optional), such as for some case, they don't want to deploy the karmada-apiserver and etcd.

Got it.